lib/model/pg-migrator.js

```diff
@@ -0,0 +1,189 @@
+      const maxLen = 255;
+      await client.query(`
+        CREATE TABLE IF NOT EXISTS post_knex_migrations (
+          id             SERIAL                      PRIMARY KEY,
+          name           VARCHAR(${maxLen})          NOT NULL,
+          batch          INTEGER                     NOT NULL,
+          migration_time TIMESTAMP(3) WITH TIME ZONE NOT NULL
+      )`);
+      log('Table now definitely exists.');
+
+      log('Acquiring lock on post_knex_migrations table...');
+      await client.query('LOCK TABLE post_knex_migrations IN EXCLUSIVE MODE NOWAIT');
+      log('Lock acquired.');
+
+      const toRun = await getMigrationsToRun(client);
+
+      if (!toRun.length) {
+        log('No migrations to run - exiting.');
+        await client.query('ROLLBACK');
+        return;
+      }
+
+      log('Validating', toRun.length, 'migrations...');
+      for (const { migration, name } of toRun) {
+        log('Validing migration:', name, '...');
+
+        if (name.length > maxLen) throw new Error(`Migration name '${name}' is too long - max length is ${maxLen}, but got ${name.length}`);
+
+        // TODO check for illegal chars in name?
+
+        const keys = Object.keys(migration);
+        const unexpectedKeys = _.omit(keys, 'up', 'down');
+        if (unexpectedKeys.length) throw new Error(`Unexpected key(s) found in migration ${name}: ${unexpectedKeys}`);
+
+        if (!migration.up) throw new Error(`Required prop .up not found in migration ${name}`);
+        if (typeof migration.up !== 'function') {
+          throw new Error(`Required prop .up of migration ${name} has incorrect type - expected 'function', but got '${typeof migration.up}'`);
+        }
+
+        if (migration.down && typeof migration.down !== 'function') {
+          throw new Error(`Optional prop .down of migration ${name} has incorrect type - expected 'function' but got '${typeof migration.down}'`);
+        }
+
+        log('Migration', name, 'looks valid.');
+      }
+      log(toRun.length, 'migrations look valid.');
+
+      log('Running', toRun.length, 'migrations...');
+      for (const { migration, name } of toRun) {
+        log('Running migration:', name);
+        await migration.up(client); // eslint-disable-line no-await-in-loop
+        log('Migration complete:', name);
+      }
+      log(toRun.length, 'migrations ran OK.');
+
+      const { lastBatch } = (await client.query(`SELECT COALESCE(MAX(batch), 0) AS "lastBatch" FROM post_knex_migrations`)).rows[0];
+      log('lastBatch:', lastBatch);
+
+      // Note that migration_time is CLOCK_TIMESTAMP() to match knex implementation.
+      // TODO confirm in relevant version of knex source code that this is actually the case, and link here.
+      const namesJson = JSON.stringify(toRun.map(m => m.name));
+      // See: https://www.postgresql.org/docs/current/functions-json.html
+      await client.query(`
+        INSERT INTO post_knex_migrations(name, batch, migration_time)
```

Comment:
Use the knex tables
* change this table name to knex_
* add the other knex lock table